### PR TITLE
add architecture support for mips(64 bit mainframe)

### PR DESCRIPTION
**Pull request by sunny868**
The patches are available in the "loongon_criu-dev-v1" branch under:

 * https://github.com/sunny868/criu/tree/loongson-criu-dev-v1

### CODE REVIEW

**Comment by sunny868** [comment]
I look forward to the feedback from the maintainer.  ^_^

File: compel/arch/mips/plugins/std/syscalls/syscall-common-mips-64.S [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	END(name)
+	
+	.text
+	.align	4
+
+
+ENTRY(__cr_restore_rt)
+	li v0,__NR_rt_sigreturn
+	syscall
+END(__cr_restore_rt)
```
Comment (0x7f454c46) [inline compel/arch/mips/plugins/std/syscalls/syscall-common-mips-64.S]: This and the definition of `SA_RESTORER` seem to be a dead-code.
Looking in mips kernel:
```
 * SA_RESTORER used to be defined as 0x04000000 but only the O32 ABI ever
 * supported its use and no libc was using it, so the entire sa-restorer
 * functionality was removed with lmo commit 39bffc12c3580ab for 2.5.48
 * retaining only the SA_RESTORER definition as a reminder to avoid
 * accidental reuse of the mask bit.
```

**Comment by sunny868** [comment]
@0x7f454c46 Sorry I have-not sent the patch for mremap()ing vDSO to lkml,but I will sent it as soon as possible.

**Comment by 0x7f454c46** [comment]
I'll review it next week - sorry for the delay caused by NY/Christmas celebrations.

File: compel/arch/mips/scripts/compel-pack-compat.lds.S [hunk 2]
```diff
@@ -0,0 +1,41 @@
+OUTPUT_ARCH(i386)
+TARGET(elf32-i386)
```
Comment (rppt) [inline compel/arch/mips/scripts/compel-pack-compat.lds.S]: I believe this is a stale copy from x86

File: compel/arch/mips/src/lib/include/uapi/asm/fpu.h [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+struct fpx_sw_bytes {
+	uint32_t			magic1;
+	uint32_t			extended_size;
+	uint64_t			xstate_bv;
+	uint32_t			xstate_size;
+	uint32_t			padding[7];
+};
+
+struct i387_fxsave_struct {
```
Comment (rppt) [inline compel/arch/mips/src/lib/include/uapi/asm/fpu.h]: This does not look like a MIPS ;-)
The entire fpu.h also seems like a stale copy from x86

File: include/common/arch/mips/asm/compiler-gcc.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
+#ifndef __LINUX_COMPILER_H
+#error "Please don't include <linux/compiler-gcc.h> directly, include <linux/compiler.h> instead."
```
Comment (rppt) [inline include/common/arch/mips/asm/compiler-gcc.h]: This file does not seem to be included anywhere

File: include/common/arch/mips/asm/regdef.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
+#ifndef __ASM_REGDEF_H
+#define __ASM_REGDEF_H
+
```
Comment (rppt) [inline include/common/arch/mips/asm/regdef.h]: The register definitions are repeated at least 3 places. Can you please use a single include file everywhere?

File: include/common/arch/mips/asm/utils.h [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+#endif
+
+typedef struct {
+	int counter;
+}atomic_t;
+
+/*from kernel/linux-3.10.84/arch/mips/include/barrier.h*/
+#if defined(CONFIG_WEAK_ORDERING) && defined(CONFIG_SMP)
+# ifdef CONFIG_CPU_CAVIUM_OCTEON
+#  define smp_mb()	__sync()
```
Comment (rppt) [inline include/common/arch/mips/asm/utils.h]: I think all the ifdefery with kernel config options can be removed as we cannot rely on them in userspace.
I'd choose the most restrictive implementation of the barrier primitives so that they would work correctly on any MIPS variant even at expense of performance

File: compel/Makefile [hunk 7]
```diff
@@ -17,6 +17,10 @@ ldflags-y		+= -r
 # UAPI inclusion, referred as <compel/...>
 ccflags-y		+= -I compel/include/uapi
 
+ifeq ($(ARCH),mips)
+ccflags-y		+= -I compel
```
Comment (rppt) [inline compel/Makefile]: Why is this required? Does MIPS port has different headers layout than other architectures?

File: compel/plugins/Makefile [hunk 8]
```diff
@@ -16,7 +16,12 @@ asflags-y		+= -I compel/include/uapi
 
 # General compel includes
 ccflags-y		+= -iquote compel/include
+
+ifeq ($(ARCH),mips)
+ccflags-y		+= -mno-abicalls -fno-stack-protector
+else
 ccflags-y		+= -fpie -fno-stack-protector
+endif
```
Comment (rppt) [inline compel/plugins/Makefile]: If it is possible to build MIPS variant with `-fpie`, please only add `-mno-abicalls` for MIPS, e.g. something like:
```
ccflags-y		+= -fpie -fno-stack-protector
ifeq ($(ARCH),mips)
ccflags-y		+= -mno-abicalls
endif
```

File: compel/plugins/Makefile [hunk 9]
```diff
@@ -28,7 +33,12 @@ asflags-y		+= -iquote $(PLUGIN_ARCH_DIR)/include
 asflags-y		+= -iquote $(PLUGIN_ARCH_DIR)
 
 # General flags for assembly
+ifeq ($(ARCH),mips)
+asflags-y		+= -mno-abicalls -Wstrict-prototypes
+else
 asflags-y		+= -fpie -Wstrict-prototypes
+endif
+
```
Comment (rppt) [inline compel/plugins/Makefile]: Ditto

**Comment by rppt** [comment]
@sunny868 great work!

It would be nice if you could merge the last two commits (["mips:fixbug ARCH_SI_TRAP"](https://github.com/checkpoint-restore/criu/pull/894/commits/29189af91990f2d26048c8645f3cb61f7f0359c0)" and "[mips:fixbug R_MIPS_26 reloation ](https://github.com/checkpoint-restore/criu/pull/894/commits/74f909e5601f4784d0291935f05699d2f49c6c34)") into the original commits that should be fixed.

**Comment by sunny868** [comment]
@0x7f454c46 I have sent the patch for mremap()ing vDSO to lkml in 2020/1/2 ，but didn't receive any feedback now. 
For SA_RESTORER ，I will resolve it.now criu can not support multi-thread dumpee.

**Comment by 0x7f454c46** [comment]
>  I have sent the patch for mremap()ing vDSO to lkml in 2020/1/2 ，but didn't receive any feedback now.

@sunny868 How did you send it? I can't find it in any linux mailing list..

**Comment by sunny868** [comment]
@ 0x7f454c46 
I get code by git clone https://git.kernel.org/pub/scm/linux/kernel/git/mips/linux.git 
then git checkout origin/mips-next
then get maintainers list by ./scripts/get_maintainer.pl 0001-mips-vdso-Support-mremap-for-vDSO.patch
Ralf Baechle <ralf@linux-mips.org> (supporter:MIPS)
Paul Burton <paulburton@kernel.org> (supporter:MIPS,commit_signer:1/2=50%)
James Hogan <jhogan@kernel.org> (supporter:MIPS)
Vincenzo Frascino <vincenzo.frascino@arm.com> (commit_signer:1/2=50%,...)
Allison Randal <allison@lohutok.net> (commit_signer:1/2=50%,commit_signer:1/1=100%)
Thomas Gleixner <tglx@linutronix.de> (commit_signer:1/2=50%,a...)
Greg Kroah-Hartman <gregkh@linuxfoundation.org> (commit_signer:1/2=50%...)
linux-mips@vger.kernel.org (open list:MIPS)
linux-kernel@vger.kernel.org (open list)

then git send-email 0001-mips-vdso-Support-mremap-for-vDSO.patch --to maintainer list

**Comment by sunny868** [comment]
@ 0x7f454c46 My email is sunguoyun@loongson.cn

**Comment by shrkamat** [comment]
Will this PR work for MIPS 32 bit CPU too ? If not can you please elaborate on the TODO(s).

**Comment by sunny868** [comment]
@shrkamat this PR not support MIPS 32. can I  create a TODO file in compel/arch/mips/ to elaborate it ？

**Comment by 0x7f454c46** [comment]
@sunny868 I see only "[PATCH] MIPS: Add get_thread_area syscall" and next revision in lkml.
Do you have a message-id or link to the email in lkml archive?

**Comment by sunny868** [comment]
@ 0x7f454c46 I am sorry,I forgot to CC open list before. I have sent the patch for mremap()ing vDSO to lkml again. so you can find in https://lore.kernel.org/linux-mips/1578978808-18633-1-git-send-email-sunguoyun@loongson.cn/

**Comment by 0x7f454c46** [comment]
@sunny868 that should be better, yep.

Let me review the CRIU part now..
+Cc: @avagin, @cyrillos, @rst0git, guys, can we have some more eyes on this? @sunny868 did good work and the revision seems to progress fast, let's spend a bit more time on it and finally merge it when it's ready.

File: include/common/arch/mips/asm/utils.h [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+# define kernel_uses_llsc	0
+#endif
+
+typedef struct {
+	int counter;
+}atomic_t;
+
+
+#if defined(CONFIG_WEAK_REORDERING_BEYOND_LLSC) && defined(CONFIG_SMP)
+#define __WEAK_LLSC_MB		"	sync	\n"
```
Comment (0x7f454c46) [inline include/common/arch/mips/asm/utils.h]: It shouldn't be compile-time option - instead, consider it's always true and add
```
/*
 * FIXME: detect with compel_cpu_has_feature() if LL/SC implicitly
 * provide a memory barrier.
 */
```

File: compel/arch/mips/plugins/std/sys-exec-tbl-64.c [hunk 11]
```diff
@@ -0,0 +1,106 @@
+/* Autogenerated, don't edit */
```
Comment (0x7f454c46) [inline compel/arch/mips/plugins/std/sys-exec-tbl-64.c]: The file is build-time generated - we don't need it in git. Drop it please.

File: compel/arch/mips/plugins/std/syscalls-64.S [hunk 12]
```diff
@@ -0,0 +1,108 @@
+/* Autogenerated, don't edit */
```
Comment (0x7f454c46) [inline compel/arch/mips/plugins/std/syscalls-64.S]: The same, drop the file, please.

File: compel/src/lib/handle-elf.c [hunk 13]
```diff
@@ -405,6 +406,114 @@ int __handle_elf(void *mem, size_t size)
 #endif
 
 			switch (ELF_R_TYPE(r->rel.r_info)) {
+#ifdef ELF_MIPS
+	      /*fixme:gysun
```
Comment (0x7f454c46) [inline compel/src/lib/handle-elf.c]: what's the fixme about?

File: compel/src/lib/handle-elf.c [hunk 14]
```diff
@@ -405,6 +406,114 @@ int __handle_elf(void *mem, size_t size)
...
+				break;
+			case R_MIPS_GOT_OFST:
+			      /*TODO*/
+			    pr_debug("\t\t\t\tTODO:R_MIPS_GOT_OFST      \n");
+				break;
+			case R_MIPS_JALR:
+			      /*TODO*/
+			    pr_debug("\t\t\t\tTODO:R_MIPS_JALR      \n");
+				break;
+#endif //0
```
Comment (0x7f454c46) [inline compel/src/lib/handle-elf.c]: Please, don't add relocations those are in TODO: otherwise, they get ignored and it may lead to hard-to-debug crashes in PIEs.

File: criu/arch/mips/crtools.c [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+void arch_free_thread_info(CoreEntry *core)
+{
+	if (!core->thread_info)
+		return;
+
+	if (core->thread_info->fpregs->xsave)
+		xfree(core->thread_info->fpregs->xsave->ymmh_space);
+	xfree(core->thread_info->fpregs->st_space);
+	xfree(core->thread_info->fpregs->xmm_space);
+	xfree(core->thread_info);
```
Comment (0x7f454c46) [inline criu/arch/mips/crtools.c]: `core->thread_info` is x86-specific, use `ti_mips` here

File: criu/arch/mips/crtools.c [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+
+    f->rs_uc.uc_mcontext.sc_mdlo = r->lo;
+    f->rs_uc.uc_mcontext.sc_mdhi = r->hi;
+    f->rs_uc.uc_mcontext.sc_pc = r->cp0_epc;
+    return 0;
+}
+
+int get_task_futex_robust_list_compat(pid_t pid, ThreadCoreEntry *info)
+{
+    pr_warn("%s:%d:no implemented!!\n",__FILE__,__LINE__);
```
Comment (0x7f454c46) [inline criu/arch/mips/crtools.c]: no need for warning as it'll be seen every C/R

File: criu/arch/mips/include/asm/compat.h [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+#ifdef CONFIG_COMPAT
+extern unsigned long call32_from_64(void *stack, void *func);
+#endif
+
+#ifndef CR_NOGLIBC
+# undef sys_mmap
+# undef sys_munmap
+#endif
+
+#endif
```
Comment (0x7f454c46) [inline criu/arch/mips/include/asm/compat.h]: Please, drop this file as it's x86-specific.

File: criu/arch/mips/include/asm/dump.h [hunk 18]
```diff
@@ -0,0 +1,15 @@
+#ifndef __CR_ASM_DUMP_H__
+#define __CR_ASM_DUMP_H__
+
+extern int save_task_regs(void *, user_regs_struct_t *, user_fpregs_struct_t *);
+extern int arch_alloc_thread_info(CoreEntry *core);
+extern void arch_free_thread_info(CoreEntry *core);
+extern int get_task_futex_robust_list_compat(pid_t pid, ThreadCoreEntry *info);
+
+/*fixme: gysun*/
```
Comment (0x7f454c46) [inline criu/arch/mips/include/asm/dump.h]: What is the fixme about?

File: criu/arch/mips/include/asm/parasite.h [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+    asm("rdhwr %0, $29" : "=r"(*ptls));
+
+    	/* asm volatile(							 */
+	/* 	     "move $4, %0				    \n"	 */
+	/* 	     "li $2,  "__stringify(__NR_get_thread_area)"  \n"  */
+	/* 	     "syscall					    \n"	 */
+	/* 	     :							 */
+	/* 	     : "r"(ptls)					 */
+	/* 	     : "$4","$2","memory"); */
+
```
Comment (0x7f454c46) [inline criu/arch/mips/include/asm/parasite.h]: Just remove this comment as it shouldn't work on mips.

File: criu/arch/mips/include/asm/restore.h [hunk 20]
```diff
@@ -0,0 +1,33 @@
+#ifndef __CR_ASM_RESTORE_H__
+#define __CR_ASM_RESTORE_H__
+
+#include "asm/restorer.h"
+
+#include "images/core.pb-c.h"
+
+/*fixme: gysun*/
```
Comment (0x7f454c46) [inline criu/arch/mips/include/asm/restore.h]: What is this fixme about?

File: criu/arch/mips/include/asm/restore.h [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+		     "jalr $25   				    \n"		\
+		     "nop   				    \n"		\
+		     :							\
+		     :"r"(task_args),"r"(restore_task_exec_start),	\
+		      "g"(new_sp)			\
+		     : "$29", "$25", "$4","$5")
+
+static inline void core_get_tls(CoreEntry *pcore, tls_t *ptls)
+{
+      /*fixme: gysun */
```
Comment (0x7f454c46) [inline criu/arch/mips/include/asm/restore.h]: What is this fixme about?

File: criu/arch/mips/include/asm/restorer.h [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+#include <compel/plugins/std/syscall-codes.h>
+#include <compel/asm/sigframe.h>
+#include "asm/compat.h"
+
+#ifdef CONFIG_COMPAT
+extern void restore_tls(tls_t *ptls);
+extern int arch_compat_rt_sigaction(void *stack32, int sig,
+		rt_sigaction_t_compat *act);
+extern int set_compat_robust_list(uint32_t head_ptr, uint32_t len);
+#else /* CONFIG_COMPAT */
```
Comment (0x7f454c46) [inline criu/arch/mips/include/asm/restorer.h]: Remove those, please

File: criu/arch/mips/include/asm/restorer.h [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+		     :  "memory")
+
+#define ARCH_FAIL_CORE_RESTORE					/* \ */
+	/* asm volatile(						\ */
+	/* 	     "movq %0, %%rsp			    \n"	\ */
+	/* 	     "movq 0, %%rax			    \n"	\ */
+	/* 	     "jmp *%%rax			    \n"	\ */
+	/* 	     :						\ */
+	/* 	     : "r"(ret)					\ */
+	/* 	     : "memory") */
```
Comment (0x7f454c46) [inline criu/arch/mips/include/asm/restorer.h]: Don't define this as it's nowhere used.
I'll submit patch to remove `ARCH_FAIL_CORE_RESTORE` from other architectures.

File: criu/arch/mips/include/asm/restorer.h [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+
+#ifndef ARCH_MAP_VDSO_32
+# define ARCH_MAP_VDSO_32		0x2002
+#endif
+
+#ifndef ARCH_MAP_VDSO_64
+# define ARCH_MAP_VDSO_64		0x2003
+#endif
```
Comment (0x7f454c46) [inline criu/arch/mips/include/asm/restorer.h]: it's x86-specific for `arch_prctl()`, remove those.

File: criu/arch/mips/include/asm/restorer.h [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+#define arch_map_vdso(map, compat)		-1
+
+
+int restore_gpregs(struct rt_sigframe *f, UserMipsRegsEntry *r);
+int restore_nonsigframe_gpregs(UserMipsRegsEntry *r);
+
+int ptrace_set_breakpoint(pid_t pid, void *addr);
+int ptrace_flush_breakpoints(pid_t pid);
+
+//extern int arch_map_vdso(unsigned long map_at, bool compatible);
```
Comment (0x7f454c46) [inline criu/arch/mips/include/asm/restorer.h]: no reason to keep this comment

File: criu/arch/mips/include/asm/types.h [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+
+#define AT_VECTOR_SIZE			44
+typedef uint64_t auxv_t;
+
+/*
+ */
+#define GDT_ENTRY_TLS_MIN		12
+#define GDT_ENTRY_TLS_MAX		14
+#define GDT_ENTRY_TLS_NUM		3
+/*fixme :gysun*/
```
Comment (0x7f454c46) [inline criu/arch/mips/include/asm/types.h]: those are x86-specific, please remove them

File: criu/arch/mips/include/asm/vdso.h [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+#define VDSO_SYMBOL_MAX	7
+#define VDSO_SYMBOL_GTOD	0
+#define ARCH_VDSO_SYMBOLS			\
+	"__vdso_clock_gettime",			\
+	"__vdso_getcpu",			\
+	"__vdso_gettimeofday",			\
+	"__vdso_time",				\
+	"__kernel_vsyscall",			\
+	"__kernel_sigreturn",			\
+	"__kernel_rt_sigreturn"
```
Comment (0x7f454c46) [inline criu/arch/mips/include/asm/vdso.h]: kernel support for mips is:
```
		__vdso_clock_gettime;
		__vdso_gettimeofday;
		__vdso_clock_getres;
```

File: criu/arch/mips/restorer.c [hunk 28]
```diff
@@ -0,0 +1,60 @@
+/*fixme: gysun */
+//#include <asm/prctl.h>
```
Comment (0x7f454c46) [inline criu/arch/mips/restorer.c]: please, remove those while unused

File: criu/arch/mips/restorer.c [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+void restore_tls(tls_t *ptls)
+{
+	asm volatile(							
+		     "move $4, %0				    \n"	
+		     "li $2,  "__stringify(__NR_set_thread_area)"  \n" 
+		     "syscall					    \n"	
+		     :							
+		     : "r"(*ptls)					
+		     : "$4","$2","memory");
+}
```
Comment (0x7f454c46) [inline criu/arch/mips/restorer.c]: please, remove CONFIG_COMPAT-related parts.
you can bring it back later if you plan to introduce compatible applications C/R support

File: criu/arch/mips/restorer_unmap.S [hunk 30]
```diff
@@ -0,0 +1,13 @@
...
+	.text
+ENTRY(__export_unmap_compat)
+	;; .code32
+	;; mov	bootstrap_start, %ebx
+	;; mov	bootstrap_len, %ecx
+	;; sub	vdso_rt_size, %ecx
+	;; movl	$__NR32_munmap, %eax
+	;; int	$0x80
+	;; int	$0x03 /* Guard */
+	.code64
```
Comment (0x7f454c46) [inline criu/arch/mips/restorer_unmap.S]: ditto, please, remove

File: criu/arch/mips/sigaction_compat_pie.c [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+
+extern char restore_rt_sigaction;
+
+/*
+ */
+int arch_compat_rt_sigaction(void *stack32, int sig, rt_sigaction_t_compat *act)
+{
+    return 0;
+}
+
```
Comment (0x7f454c46) [inline criu/arch/mips/sigaction_compat_pie.c]: ditto, please, remove

File: criu/arch/mips/vdso-pie.c [hunk 32]
```diff
@@ -0,0 +1,76 @@
...
+		u32	guards;
+	} __packed jmp = {
+		.movl		= 0xb8,
+		.imm32		= (uint32_t)to,
+		.jmp_eax	= 0xe0ff,
+		.guards		= 0xcccccccc,
+	};
+
+	memcpy((void *)from, &jmp, sizeof(jmp));
+}
```
Comment (0x7f454c46) [inline criu/arch/mips/vdso-pie.c]: please, remove 32-bit part for now

File: criu/arch/mips/vdso-pie.c [hunk 33]
```diff
@@ -0,0 +1,76 @@
...
+		u16	movabs;
+		u64	imm64;
+		u16	jmp_rax;
+		u32	guards;
+	} __packed jmp = {
+		.movabs		= 0xb848,
+		.imm64		= to,
+		.jmp_rax	= 0xe0ff,
+		.guards		= 0xcccccccc,
+	};
```
Comment (0x7f454c46) [inline criu/arch/mips/vdso-pie.c]: you need to adjust the code for mips64 assembly here.
For more information, what's this about:
https://criu.org/Vdso

File: Makefile [hunk 34]
```diff
@@ -76,6 +76,10 @@ ifeq ($(ARCH),x86)
         DEFINES		:= -DCONFIG_X86_64
 endif
 
+ifeq ($(ARCH),mips)
+        DEFINES		:= -DCONFIG_MIPS -DCONFIG_SMP  -DCONFIG_WEAK_REORDERING_BEYOND_LLSC
```
Comment (0x7f454c46) [inline Makefile]: please, remove `-DCONFIG_SMP  -DCONFIG_WEAK_REORDERING_BEYOND_LLSC` as suggested for earlier patches

File: Makefile [hunk 35]
```diff
@@ -105,6 +109,10 @@ WARNINGS		:= -Wall -Wformat-security
 CFLAGS-GCOV		:= --coverage -fno-exceptions -fno-inline -fprofile-update=atomic
 export CFLAGS-GCOV
 
+ifeq ($(ARCH),mips)
+WARNINGS		:= -g -rdynamic
+endif
+
```
Comment (0x7f454c46) [inline Makefile]: this doesn't seem to be mips-specific, rather turned on by `DEBUG=1 make`

File: images/core.proto [hunk 36]
```diff
@@ -93,6 +95,7 @@ message thread_core_entry {
 	optional uint32			seccomp_filter	= 12;
 
 	optional string			comm		= 13;
+	optional uint64			blk_sigset_mips	= 14;
```
Comment (0x7f454c46) [inline images/core.proto]: Quoting from [protobuf2 specification](https://developers.google.com/protocol-buffers/docs/proto#updating):

> `optional` is compatible with `repeated`. Given serialized data of a repeated field as input, clients that expect this field to be `optional` will take the last input value if it's a primitive type field or merge all input elements if it's a message type field. 

So, I believe it's possible to change:
```
	optional uint64			blk_sigset	= 6;
```
into
```
	repeated uint64			blk_sigset	= 6;
```
and adjust the code.

File: images/core.proto [hunk 37]
```diff
@@ -53,6 +54,7 @@ message task_core_entry {
 	//optional int32		tty_pgrp	= 17;
 
 	optional bool			child_subreaper	= 18;
+	optional uint64			blk_sigset_mips = 16[(criu).hex = true];
```
Comment (0x7f454c46) [inline images/core.proto]: Let's rename it into `blk_sigset_extended` - in case there're other architectures with a big sigset.

File: criu/cr-restore.c [hunk 38]
```diff
@@ -455,9 +455,16 @@ static int restore_native_sigaction(int sig, SaEntry *e)
...
+	e->has_mask_mips = 1;
+	BUILD_BUG_ON(sizeof(e->mask)* 2 != sizeof(act.rt_sa_mask.sig));
+
+	memcpy(&(act.rt_sa_mask.sig[0]), &e->mask, sizeof(act.rt_sa_mask.sig[0]));
+	memcpy(&(act.rt_sa_mask.sig[1]), &e->mask_mips, sizeof(act.rt_sa_mask.sig[1]));
+#else
 	BUILD_BUG_ON(sizeof(e->mask) != sizeof(act.rt_sa_mask.sig));
 	memcpy(act.rt_sa_mask.sig, &e->mask, sizeof(act.rt_sa_mask.sig));
-
+#endif
```
Comment (0x7f454c46) [inline criu/cr-restore.c]: Instead of ifdeffing, separate into a helper-function and check if the `optional` field is present.

File: criu/parasite-syscall.c [hunk 39]
```diff
@@ -181,7 +181,14 @@ int parasite_dump_thread_seized(struct parasite_thread_ctl *tctl,
...
 
 	tc->has_blk_sigset = true;
+#ifdef CONFIG_MIPS
+
+	memcpy(&tc->blk_sigset, compel_thread_sigmask(tctl), sizeof(tc->blk_sigset));
+	memcpy(&tc->blk_sigset_mips, compel_thread_sigmask(tctl)+1, sizeof(tc->blk_sigset));
+#else
+
 	memcpy(&tc->blk_sigset, compel_thread_sigmask(tctl), sizeof(k_rtsigset_t));
+#endif
```
Comment (0x7f454c46) [inline criu/parasite-syscall.c]: ditto

File: criu/parasite-syscall.c [hunk 40]
```diff
@@ -242,8 +249,16 @@ int parasite_dump_sigacts_seized(struct parasite_ctl *ctl, struct pstree_item *i
...
+#ifdef CONFIG_MIPS
+		sa->has_mask_mips = 1;
+		BUILD_BUG_ON(sizeof(sa->mask) * 2 != sizeof(args->sas[0].rt_sa_mask.sig));
+		memcpy(&sa->mask, &(args->sas[i].rt_sa_mask.sig[0]), sizeof(sa->mask));
+		memcpy(&sa->mask_mips, &(args->sas[i].rt_sa_mask.sig[1]), sizeof(sa->mask));
+
+#else
 		BUILD_BUG_ON(sizeof(sa->mask) != sizeof(args->sas[0].rt_sa_mask.sig));
 		memcpy(&sa->mask, args->sas[i].rt_sa_mask.sig, sizeof(sa->mask));
+#endif
```
Comment (0x7f454c46) [inline criu/parasite-syscall.c]: ditto

File: criu/parasite-syscall.c [hunk 41]
```diff
@@ -571,7 +586,14 @@ struct parasite_ctl *parasite_infect_seized(pid_t pid, struct pstree_item *item,
...
 
 	parasite_args_size = PARASITE_ARG_SIZE_MIN; /* reset for next task */
+#ifdef CONFIG_MIPS
+
+	memcpy(&item->core[0]->tc->blk_sigset, (unsigned long *)compel_task_sigmask(ctl), sizeof(item->core[0]->tc->blk_sigset));
+	memcpy(&item->core[0]->tc->blk_sigset_mips, (unsigned long *)compel_task_sigmask(ctl)+1, sizeof(item->core[0]->tc->blk_sigset));
+
+#else
 	memcpy(&item->core[0]->tc->blk_sigset, compel_task_sigmask(ctl), sizeof(k_rtsigset_t));
+#endif
```
Comment (0x7f454c46) [inline criu/parasite-syscall.c]: ditto

**Comment by 0x7f454c46** [comment]
On top of review comments, please do two other things:
1. Try compiling and run zdtm tests to see what doesn't work; post a link to gist with results in the PR.
2. Add another commit on the top that introduces a cross-compilation job on Travis-CI for mips64 (into `.travis.yml`, see as an example `armv7-cross` job). This way CRIU contributors can verify that with their changes CRIU _at least_ still builds for mips architecture.
Thanks, nice work!

File: compel/src/lib/handle-elf.c (hunk 14 above)
Comment (0x7f454c46) [reply compel/src/lib/handle-elf.c]: (so either drop those or put the code to handle them)

**Comment by sunny868** [comment]
@0x7f454c46 Thank you very much for checking my code so carefully. I'm currently on Chinese New Year's holiday. I'll fix the code and test it as soon as possible later.

**Comment by shrkamat** [comment]
> @shrkamat this PR not support MIPS 32. can I create a TODO file in compel/arch/mips/ to elaborate it ？

Yes it should help

**Comment by sunny868** [comment]
@0x7f454c46 @rppt @shrkamat @adrianreber  I am sorry I have create a new PR (#933 )again. this PR #894 fork is too old, resulting in many conflicts.

File: compel/Makefile (hunk 7 above)
Comment (sunny868) [reply compel/Makefile]: it is used to compile handler-elf.c. in compel/src/lib/handler-elf.c . I used this :
#ifdef CONFIG_MIPS
#include "arch/mips/src/lib/include/ldsodefs.h"
#endif

File: compel/plugins/Makefile (hunk 8 above)
Comment (sunny868) [reply compel/plugins/Makefile]: MIPS can not use -fpie or -fpic when used -mno-abicalls

**Comment by adrianreber** [comment]
Closing this PR as MIPS support has been merged via #933
